File: nodejs-mvc/hexagonal-architecture/src/domain/service/ProductService.ts
import { Product } from '../model/Product';
import { ProductRepositoryPort } from '../port/output/ProductRepositoryPort';
import {
  CreateProductCommand,
  PaginatedResult,
  ProductServicePort,
  SearchCriteria,
  UpdateProductCommand,
} from '../port/input/ProductServicePort';
import { v4 as uuidv4 } from 'uuid';

/**
 * @layer Domain Service
 * @description Implements the input port (ProductServicePort) and orchestrates
 * domain logic. This is the APPLICATION SERVICE in Hexagonal Architecture.
 *
 * It depends on:
 * - The domain model (Product)
 * - The output port interface (ProductRepositoryPort)
 *
 * It does NOT depend on:
 * - Express or any web framework
 * - Any specific database or persistence technology
 * - Any external infrastructure
 */
export class ProductService implements ProductServicePort {
  constructor(private readonly repository: ProductRepositoryPort) {}

  async createProduct(command: CreateProductCommand): Promise<Product> {
    // Business rule: price must be > 0
    Product.validatePrice(command.price);

    // Business rule: stock cannot be negative
    Product.validateStock(command.stockQuantity);

    // Business rule: SKU must be unique
    const existingProduct = await this.repository.findBySku(command.sku);
    if (existingProduct) {
      throw new Error(`Product with SKU '${command.sku}' already exists`);
    }

    const now = new Date();
    const product = new Product({
      id: uuidv4(),
      name: command.name,
      description: command.description,
      price: command.price,
      category: command.category,
      stockQuantity: command.stockQuantity,
      sku: command.sku,
      active: true,
      createdAt: now,
      updatedAt: now,
    });

    return this.repository.save(product);
  }

  async getProductById(id: string): Promise<Product> {
    const product = await this.repository.findById(id);
    if (!product) {
      throw new Error(`Product with id '${id}' not found`);
    }
    return product;
  }

  async listProducts(page: number, size: number): Promise<PaginatedResult<Product>> {
    // Business rule: max page size is 100
    const effectiveSize = Math.min(size, 100);
    const effectivePage = Math.max(page, 0);

    const { products, total } = await this.repository.findAll(effectivePage, effectiveSize);

    return {
      content: products,
      page: effectivePage,
      size: effectiveSize,
      totalElements: total,
      totalPages: Math.ceil(total / effectiveSize),
    };
  }

  async searchProducts(criteria: SearchCriteria): Promise<Product[]> {
    return this.repository.search(
      criteria.query,
      criteria.category,
      criteria.minPrice,
      criteria.maxPrice
    );
  }

  async updateProduct(id: string, command: UpdateProductCommand): Promise<Product> {
    const existing = await this.getProductById(id);

    if (command.price !== undefined) {
      Product.validatePrice(command.price);
    }

    if (command.stockQuantity !== undefined) {
      Product.validateStock(command.stockQuantity);
    }

    // Check SKU uniqueness if changing SKU
    if (command.sku && command.sku !== existing.sku) {
      const existingBySku = await this.repository.findBySku(command.sku);
      if (existingBySku) {
        throw new Error(`Product with SKU '${command.sku}' already exists`);
      }
    }

    const updatedProduct = new Product({
      id: existing.id,
      name: command.name ?? existing.name,
      description: command.description ?? existing.description,
      price: command.price ?? existing.price,
      category: command.category ?? existing.category,
      stockQuantity: command.stockQuantity ?? existing.stockQuantity,
      sku: command.sku ?? existing.sku,
      active: existing.active,
      createdAt: existing.createdAt,
      updatedAt: new Date(),
    });

    return this.repository.update(updatedProduct);
  }

  async deleteProduct(id: string): Promise<void> {
    const product = await this.getProductById(id);
    const deactivated = product.deactivate();
    await this.repository.update(deactivated);
  }

  async decreaseStock(id: string, quantity: number): Promise<Product> {
    const product = await this.getProductById(id);
    const updated = product.decreaseStock(quantity);
    return this.repository.update(updated);
  }

  async increaseStock(id: string, quantity: number): Promise<Product> {
    const product = await this.getProductById(id);
    const updated = product.increaseStock(quantity);
    return this.repository.update(updated);
  }
}

File: python-mvc/hexagonal-architecture/app/domain/service/product_service.py
"""
Domain Service: Product Service.

This service implements the input port (ProductServicePort) and orchestrates
domain logic by coordinating between the domain model and the output port
(ProductRepositoryPort).

It contains application-level business rules such as:
- Validating uniqueness constraints (SKU)
- Coordinating persistence operations
- Enforcing pagination limits

This service depends ONLY on domain abstractions (ports), never on
concrete infrastructure implementations.
"""
from __future__ import annotations

import math
from dataclasses import replace
from datetime import datetime

from app.domain.model.product import (
    DuplicateSkuError,
    Product,
    ProductNotFoundError,
)
from app.domain.port.input.product_service_port import (
    CreateProductCommand,
    PaginatedResult,
    ProductServicePort,
    SearchCriteria,
    UpdateProductCommand,
)
from app.domain.port.output.product_repository_port import ProductRepositoryPort


class ProductService(ProductServicePort):
    """
    Implementation of the ProductServicePort.

    Orchestrates domain operations using the repository port for persistence.
    """

    MAX_PAGE_SIZE = 100

    def __init__(self, repository: ProductRepositoryPort) -> None:
        self._repository = repository

    async def create_product(self, command: CreateProductCommand) -> Product:
        """Create a new product after validating business rules."""
        # Validate business rules
        Product.validate_price(command.price)
        Product.validate_stock(command.stock_quantity)

        # Check SKU uniqueness
        existing = await self._repository.find_by_sku(command.sku)
        if existing is not None:
            raise DuplicateSkuError(f"Product with SKU '{command.sku}' already exists")

        # Create domain entity
        now = datetime.utcnow()
        product = Product(
            id=Product.generate_id(),
            name=command.name,
            description=command.description,
            price=command.price,
            category=command.category,
            stock_quantity=command.stock_quantity,
            sku=command.sku,
            active=True,
            created_at=now,
            updated_at=now,
        )

        return await self._repository.save(product)

    async def get_product_by_id(self, product_id: str) -> Product:
        """Retrieve a product by ID or raise ProductNotFoundError."""
        product = await self._repository.find_by_id(product_id)
        if product is None:
            raise ProductNotFoundError(f"Product with id '{product_id}' not found")
        return product

    async def list_products(self, page: int = 1, size: int = 10) -> PaginatedResult:
        """List products with pagination, enforcing max page size."""
        page = max(1, page)
        size = max(1, min(size, self.MAX_PAGE_SIZE))

        items, total = await self._repository.find_all(page, size)
        total_pages = math.ceil(total / size) if total > 0 else 0

        return PaginatedResult(
            items=items,
            total=total,
            page=page,
            size=size,
            total_pages=total_pages,
        )

    async def search_products(self, criteria: SearchCriteria) -> list[Product]:
        """Search products based on criteria."""
        return await self._repository.search(
            query=criteria.query,
            category=criteria.category,
            min_price=criteria.min_price,
            max_price=criteria.max_price,
        )

    async def update_product(
        self, product_id: str, command: UpdateProductCommand
    ) -> Product:
        """Update an existing product with the provided fields."""
        product = await self._repository.find_by_id(product_id)
        if product is None:
            raise ProductNotFoundError(f"Product with id '{product_id}' not found")

        # Validate new values if provided
        if command.price is not None:
            Product.validate_price(command.price)
        if command.stock_quantity is not None:
            Product.validate_stock(command.stock_quantity)

        # Check SKU uniqueness if changing SKU
        if command.sku is not None and command.sku != product.sku:
            existing = await self._repository.find_by_sku(command.sku)
            if existing is not None:
                raise DuplicateSkuError(
                    f"Product with SKU '{command.sku}' already exists"
                )

        # Build updated product using only provided fields
        updates: dict = {}
        if command.name is not None:
            updates["name"] = command.name
        if command.description is not None:
            updates["description"] = command.description
        if command.price is not None:
            updates["price"] = command.price
        if command.category is not None:
            updates["category"] = command.category
        if command.stock_quantity is not None:
            updates["stock_quantity"] = command.stock_quantity
        if command.sku is not None:
            updates["sku"] = command.sku
        if command.active is not None:
            updates["active"] = command.active

        updates["updated_at"] = datetime.utcnow()
        updated_product = replace(product, **updates)

        return await self._repository.update(updated_product)

    async def delete_product(self, product_id: str) -> None:
        """Delete a product by ID."""
        product = await self._repository.find_by_id(product_id)
        if product is None:
            raise ProductNotFoundError(f"Product with id '{product_id}' not found")
        await self._repository.delete(product_id)

    async def decrease_stock(self, product_id: str, quantity: int) -> Product:
        """Decrease stock for a product."""
        product = await self._repository.find_by_id(product_id)
        if product is None:
            raise ProductNotFoundError(f"Product with id '{product_id}' not found")

        updated_product = product.decrease_stock(quantity)
        return await self._repository.update(updated_product)

    async def increase_stock(self, product_id: str, quantity: int) -> Product:
        """Increase stock for a product."""
        product = await self._repository.find_by_id(product_id)
        if product is None:
            raise ProductNotFoundError(f"Product with id '{product_id}' not found")

        updated_product = product.increase_stock(quantity)
        return await self._repository.update(updated_product)
